docs/topics/coroutines-basics.md

```diff
@@ -3,212 +3,326 @@
+> You can display coroutine names next to thread names in the output of your code for additional information.
+> In IntelliJ IDEA, right-click the `Run` icon: ![Run icon](run-icon.png){width=30} next to your `main()` function, select `Modify Run Configuration...`, and add `-Dkotlinx.coroutines.debug` in `VM options`.
+> 
+> See [Debug coroutines using IntelliJ IDEA — Tutorial](debug-coroutines-with-idea.md) for more information.
+> 
+{style="tip"}  
 
-### Structured concurrency
+To create a coroutine in Kotlin, you need a suspending function, a scope, a builder to start it, and a dispatcher to control which threads it uses:
 
-Coroutines follow a principle of 
-**structured concurrency** which means that new coroutines can only be launched in a specific [CoroutineScope]
-which delimits the lifetime of the coroutine. The above example shows that [runBlocking] establishes the corresponding
-scope and that is why the previous example waits until `World!` is printed after a second's delay and only then exits.
+1. Import the `kotlinx.coroutines` library:
 
-In a real application, you will be launching a lot of coroutines. Structured concurrency ensures that they are not
-lost and do not leak. An outer scope cannot complete until all its children coroutines complete. 
-Structured concurrency also ensures that any errors in the code are properly reported and are never lost.  
+    ```kotlin
+    import kotlinx.coroutines.*
+    ```
 
-## Extract function refactoring
+2. Mark functions that can pause and resume with the `suspend` keyword:
 
-Let's extract the block of code inside `launch { ... }` into a separate function. When you
-perform "Extract function" refactoring on this code, you get a new function with the `suspend` modifier.
-This is your first _suspending function_. Suspending functions can be used inside coroutines
-just like regular functions, but their additional feature is that they can, in turn,
-use other suspending functions (like `delay` in this example) to _suspend_ execution of a coroutine.
+    ```kotlin
+    suspend fun greet() {
+        println("Hello world from a suspending function on thread: ${Thread.currentThread().name}")
+    }
+    
+    suspend fun main() {}
+    ```
+
+    > While you can mark the `main()` function as `suspend` in some projects, other projects or frameworks might not allow changing the `main()` function this way.
+    > In those cases, coroutine builder functions like [`runBlocking`](#runblocking) are the usual entry point for calling suspending code.
+    > 
+    {style="note"}
+
+
+3. Use the [`coroutineScope()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html) function to [create a `CoroutineScope`](#coroutine-scope-and-structured-concurrency) that groups your coroutines and controls their lifecycle:
+
+    ```kotlin
+    suspend fun main() = coroutineScope {
```

Comment:
I see that this series of examples follows the idea I've proposed of mentioning dispatchers early on, but as is, it's not quite the code that anyone would actually write. I was thinking of something like this (and could probably have been clearer about it):

```kotlin
suspend fun main() = withContext(Dispatchers.Default) {
    launch { greet() }
    println("This can run in parallel with the greeter coroutine")
}
```

This way, it's clear in which context all pieces of code inside `main` are running, including the `println`.